cpp/models/abm/location.cpp

```diff
@@ -34,210 +33,155 @@ namespace abm
+void Location::remove_person(Person& p)
 {
-    m_subpopulations.get_last_value()[size_t(s)] += delta;
-    assert(m_subpopulations.get_last_value()[size_t(s)] >= 0 && "subpopulations must be non-negative");
+    m_persons.erase(std::remove(m_persons.begin(), m_persons.end(), &p), m_persons.end());
+    for (auto&& cell : m_cells) {
+        cell.m_persons.erase(std::remove(cell.m_persons.begin(), cell.m_persons.end(), &p), cell.m_persons.end());
+    }
 }
 
-int Location::get_subpopulation(InfectionState s) const
+size_t Location::get_number_persons()
 {
-    return (int)m_subpopulations.get_last_value()[size_t(s)];
+    return m_persons.size();
 }
 
 /*
-For every location we have a transmission factor that is nomalized to m_capacity.volume / m_capacity.persons of 
-the location "Home", which is 66. We multiply this rate with the factor m_capacity.persons / m_capacity.volume of the 
-considered location. For the dependency of the transmission risk on the occupancy we use a linear approximation which 
-leads to: 66 * (m_capacity.persons / m_capacity.volume) * (m_num_persons / m_capacity.persons).
+For every cell in a location we have a transmission factor that is nomalized to m_capacity.volume / m_capacity.persons of 
+the location "Home", which is 66. We multiply this rate with the individual size of each cell to obtain a "space per person" factor.
 */
-ScalarType Location::compute_relative_transmission_risk()
+ScalarType Cell::compute_space_per_person_relative()
 {
-    if (m_capacity.volume != 0 && m_capacity_adapted_transmission_risk) {
-        return 66.0 * m_num_persons / m_capacity.volume;
+    if (m_capacity.volume != 0) {
+        return 66.0 / m_capacity.volume;
     }
     else {
         return 1.0;
     }
 }
 
-void Location::add_subpopulations_timepoint(const TimePoint& t)
+size_t Cell::get_subpopulation(TimePoint t, InfectionState state) const
+{
+    return count_if(m_persons.begin(), m_persons.end(), [&](observer_ptr<Person> p) {
+        return p->get_infection_state(t) == state;
+    });
+}
+
+size_t Location::get_subpopulation(TimePoint t, InfectionState state) const
+{
+    return count_if(m_persons.begin(), m_persons.end(), [&](observer_ptr<Person> p) {
+        return p->get_infection_state(t) == state;
+    });
+}
+
+void Location::store_subpopulations(const TimePoint t)
 {
-    // Get the previous time point index.
-    // Since index starts from 0, we need to -1 from the current time point.
-    auto last_idx = m_subpopulations.get_num_time_points() - 1;
     m_subpopulations.add_time_point(t.days());
-    m_subpopulations.get_last_value() = m_subpopulations.get_value(last_idx);
+    std::array<ScalarType, size_t(InfectionState::Count)> subpopulations{};
+    for (auto p : m_persons)
+        ++subpopulations[(size_t)p->get_infection_state(t)];
+    m_subpopulations.get_last_value() = Eigen::Map<const Eigen::VectorXd>(subpopulations.data(), subpopulations.size());
```

Comment:
could subpopulations be an Eigen Vector type so you don't need Map?

Reply:
Yes, changed.